src/psyclone/psyir/nodes/reference.py

```diff
@@ -250,6 +250,41 @@ def next_access(self):
             return all_accesses[index+1].node
         return None
 
+    def replace_symbols_using(self, table):
+        '''
+        Update any Symbols referenced by this Node with those in the
+        supplied table with matching names. If there is no match for a given
+        Symbol then it is left unchanged.
+
+        .. note:
+            Since we only use symbol names here, this won't get the correct
+            symbol if we have symbols shadowed in nested scopes, e.g.:
+
+            subroutine test
+              integer :: a
+              integer :: b = 1
+              if condition
+                ! PSyIR declares a shadowed locally-scoped a'
+                a' = 1
+                if condition2
+                  ! PSyIR declares a shadowed locally-scoped b'
+                  b' = 2
+                  a = a' + b'
```

Comment:
Thanks for adding this, but it is not specific to reference, it will happen to any node with a symbol  inside a nested scope.  Could this note be in the ScopingNode implementation or  in the developer guide new section?

Reply:
And if fact it should probably be a note regarding the copy/_refine_copy instead, since we actually conform exactly to the replace_symbols_using doctring here.

The flaw is in the copy context as this is not exactly what we need there since it changes the semantics of the resulting node tree. I would also add a issue/TODO for this copy behaviour.

Reply:
Yes, agreed. I've moved it to the appropriate place, created #2666 and updated the dev. guide too.